cin_validator/rules/cin2022_23/rule_1510.py

```diff
@@ -0,0 +1,96 @@
+from typing import Mapping
+
+import pandas as pd
+
+from cin_validator.rule_engine import rule_definition, CINTable, RuleContext
+from cin_validator.rule_engine import IssueLocator
+from cin_validator.test_engine import run_rule
+
+# Get tables and columns of interest from the CINTable object defined in rule_engine/__api.py
+# Replace ChildIdentifiers with the table name, and LAChildID with the column name you want.
+
+ChildIdentifiers = CINTable.ChildIdentifiers
+UPN = ChildIdentifiers.UPN
+
+# define characteristics of rule
+@rule_definition(
+    # write the rule code here, in place of 8500
+    code=1510,
+    # replace ChildIdentifiers with the value in the module column of the excel sheet corresponding to this rule .
+    module=CINTable.ChildIdentifiers,
+    # replace the message with the corresponding value for this rule, gotten from the excel sheet.
+    message="UPN invalid (wrong check letter at character 1)",
+    # The column names tend to be the words within the < > signs in the github issue description.
+    affected_fields=[UPN],
+)
+def validate(
+    data_container: Mapping[CINTable, pd.DataFrame], rule_context: RuleContext
+):
+    # Replace ChildIdentifiers with the name of the table you need.
+    df = data_container[ChildIdentifiers]
+
+    # implement rule logic as descriped by the Github issue. Put the description as a comment above the implementation as shown.
+    """
+    <UPN> (N00001) if present must contain the correct check letter
+
+    To calculate the check letter:
+
+    1. Multiply the individual digits by their weights as follows:
+
+    digit 2 by weight 2; digit 3 by weight 3; digit 4 by weight 4; digit 5 by weight 5; digit 6 by weight 6; digit 7 by weight 7; digit 8 by weight 8; digit 9 by weight 9; digit 10 by weight 10; digit 11 by weight 11; digit 12 by weight 12; digit 13 by weight 13.
+
+    2. Sum the individual results, divide the total by 23, and take the remainder.
+
+    3. Calculate the check letter from the result as follows:
+
+    0  = A;  1  = B;  2  = C;  3  = D;  4  = E;  5 = F;  6 = G;
+    7 = H;  8 = J;  9 = K;  10 = L;  11 = M;  12 = N;  13 = P;
+    14 = Q;  15 = R;  16 = T;  17 = U;  18 = V;  19 = W;  20 = X;
+    """
+    df.reset_index(inplace=True)
+    df2 = df[['index','UPN']]
+    df2 = df2[(df2['UPN'].str.len() == 13) & df2['UPN'].notna()]
+    df2['FIRST_CHAR'] = df2['UPN'].str[:1]
+    df2['LAST_C'] = df2['UPN'].str[-12:]
+    df2['LAST_C'] = df2['LAST_C'].apply(lambda x: int(x) if str(x).isdigit() else pd.NA)
+    df2 = df2[df2['LAST_C'].notna()]
+    for i in range(1,13):
+        df2['C' + str(i)] = (df2['UPN'].str[i].astype(int) * (i+1)) % 23
```

Comment:
You've divided by 23 twice: here and in the next line.
Check the rule definition one more time and verify if you had to do it here @PhilDeakin ?

Reply:
It doesn't make any difference as it will result in the same number.

Reply:
That's interesting. Is there a reason why you chose to do it twice then?
I'm just curious.

Reply:
The numbers in the C1-12 columns can themselves end up being bigger than 23, so if you apply the modulo there too, it will reduce the size of the integers that have to be summed.... but I doubt in the grand scheme of things it would make any difference timewise as none of the numbers are massive and addition and modulus aren't expensive functions.

Reply:
Amended it to something neater that avoids the creation of the 12 C columns.